myia/vm.py

```diff
@@ -0,0 +1,193 @@
+    """
+    Virtual Machine interface.
+
+    Attributes:
+        env: evaluation environement (currently unused)
+
+    """
+
+    def __init__(self, env) -> None:
+        """Construct a VM."""
+        self.env = env
+
+    def evaluate(self, graph: Graph, args: List[Any]):
+        """
+        Run a graph.
+
+        This will evaluate the passed-in graph and return the
+        resulting value.
+
+        """
+        frame = VMFrame(graph, args, self.env)
+        return frame.evaluate()
+
+
+class VMFrame:
+    """
+    An execution frame.
+
+    This is the main class for the VM.  It does all the bookkeeping to
+    handle the execution state.
+
+    Attributes:
+        graph: The currently executing graph.  This can change due to
+            a tail call.
+        args: Argument array for the current call.
+        values: Mapping of every node in the graph to its computed
+            value.  Currently all the values are always kept live.
+        todo: List of node that have to be visited.  Evaluation starts
+            from the end of the list.
+
+    """
+
+    class Continue(Exception):
+        """Continue step."""
+
+    class Return(Exception):
+        """End step."""
+
+    def __init__(self, graph: Graph, args: List[Any], env) -> None:
+        """Build a frame."""
+        self.env = env
+        self.values: Dict[ANFNode, Any] = dict()
+        self.reset(graph, args)
+        self.calling: VMFrame = None
+
+    def reset(self, graph: Graph, args: List[Any]) -> None:
+        """
+        Set up a call.
+
+        This is used while the frame is executing a graph to implement
+        tail calls.
+        """
+        self.graph = graph
+        self.args = args
+        self.todo: List[ANFNode] = [self.graph.return_]
+
+    def evaluate(self):
+        """Run the graph to completion and return the result."""
+        while True:
+            try:
+                self.advance()
+            except StopIteration:
+                return self.values[self.graph.return_]
+
+    def tail(self):
+        """Are we in a tail call?."""
+        return self.todo[-2] is self.graph.return_
+
+    def done(self):
+        """Are we done?."""
+        return self.graph.return_ in self.values
+
+    def get_value(self, node: ANFNode):
+        """
+        Get the value for a node.
+
+        This will check if we've already computed a value for it and
+        add it to the queue otherwise. If there was no value, we stop
+        the evaluation of the current node and restart the step, since
+        no useful work was done.
+        """
+        if node not in self.values:
+            self.todo.append(node)
+            raise self.Continue()
+        return self.values[node]
+
+    def do_call(self, graph: Graph, args: List[Any]):
+        """
+        Perform a call.
+
+        This will handle tail calls correctly by reusing the current VMFrame.
+        """
+        if self.tail():
+            self.reset(graph, args)
```

Comment:
Because (of course) of these damned closures, I don't think you can actually do that: whenever you encounter a `Constant` that contains a `Graph`, and that this graph has the current graph as a parent, you need to attach the frame to it somehow (as an env), so that it can look up its closure variables. This will sometimes entail further computation in this frame, so you can't reuse it for a different graph.

Reply:
I don't have any form of variable mapping in the frame, so I fail to see how that would help closures.

Also, if we can't reuse frames, then we can't do tail calls.

Reply:
Closures here are graphs that have pointers to nodes that belong to other graphs. The semantics at execution is that if, when evaluating graph `G`, you encounter graph `H` such that `H.parent is G`, then you must construct a closure for `H` that lets it access the values for `G`'s nodes. Each execution of `G` produces a fresh closure for `H` which is a first class value that can be passed around, and as long as that closure lives, so does `G`'s frame or whatever structure holds these values.

A tail call just means that the current frame can be popped from the stack immediately. It can still be kept alive by a closure. If no closure points to it, then the GC will eliminate it.

Reply:
I see what you mean, and I will reorganize the inner workings so that this can actually support closures correctly.  As a bonus it should waste less memory when there are no closures.